mineral-ui/mineral-ui
# [themes] Generate theme with color options

### Description

- Add `flow:watch` npm script
- Generate token files useful to themes
- Add `color_[variant]Primary` tokens
- Generate colorAliases into the themes dir, rather than tokens/src
- Refactor some of formats.js
- Release mineral-ui-tokens v0.3.0
- Generate theme values from color options
- Add `color_[variant]Primary` theme variables
- Update Avatar & Button to use `color_[variation]Primary`
- Website updates
  - Update docs for `createTheme` API
  - Update use of `createTheme` with new signature
  - Update to use `color_themePrimary`
  - Import and use palette and its type throughout
- Add specs and snapshots for createTheme & themeFromTokens
- Update snapshots after theme update

### Motivation and context

Connects #623 

### Screenshots, videos, or demo, if appropriate


https://623-theme-from-palette-2--mineral-ui.netlify.com/theming

### How to test


1. `cd mineral-ui && npm start`
2. Navigate to Theming page
3. Check the new [`createTheme` API docs](https://623-theme-from-palette-2--mineral-ui.netlify.com/theming#common-scenarios-api-{{47}}) for completeness & correctness
4. Check the [theme structure tables](https://623-theme-from-palette-2--mineral-ui.netlify.com/theming#common-scenarios-theme-structure) for completeness
5. Navigate to the [Palette Demo](https://623-theme-from-palette-2--mineral-ui.netlify.com/palette-demo)
6. Confirm you can still switch themes

### Types of changes

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

#### Breaking
- **tokens**: `color_primary` (`$mnrl-color-primary`, in Sass) token renamed to `color_themePrimary` ($mnrl-color-themePrimary)
- **themes**: Change `createTheme` signature. Replace `baseColor` and `overrides` parameters with an options object, containing `colors` (baseColor is now colors.theme) and `overrides`. See the [new documentation](http://mineral-ui.com/theming#common-scenarios-api-{{47}}) for details. Rename `color_primary` theme variable to `color_themePrimary` for consistency with other variable names

### Checklist

* [x] Rendering performance (initial load time & 60fps) and [perceived performance](http://blog.teamtreehouse.com/perceived-performance) considered (see note)

**Performance**: We optimized as much as we could, particularly createTheme and themeFromTokens, but we should consider exporting a `generateTheme` function with the same functionality of createTheme that creates a static theme.



### How does this PR make you feel?

![Theme that horse](https://media.giphy.com/media/1woAIetDOL2F5y41dG/giphy.gif)
